rustls/src/client/hs.rs

```diff
@@ -271,6 +271,47 @@ fn emit_client_hello_for_retry(
     // Do we have a SessionID or ticket cached for this host?
     let tls13_session = prepare_resumption(&input.resuming, &mut exts, suite, cx, config);
 
+    // Extensions MAY be randomised
+    // PSK message must be last, if it is we cannot move it
+    let randomisable_extensions = exts.len()
+        - if let Some(&ClientExtension::PresharedKey(..)) = exts.last() {
+            1
+        } else {
+            0
+        };
+
+    let start = input
+        .hello
+        .sent_extensions
+        .iter()
+        .filter(|&&ext_type| ext_type != ExtensionType::PreSharedKey)
+        .fold(0, |acc, &ext_type| {
+            let pos = exts
+                .iter()
+                // The initial `acc` were already ordered, so we cannot find the extension in those
+                .skip(acc)
+                .map(|ext| ext.ext_type())
+                .enumerate()
+                .find(|(_, new_type)| ext_type == *new_type);
+
+            if let Some((pos, _)) = pos {
+                // By putting all extensions that have been seen in the start of the vector we
+                // make others bubble to the end
+                exts.swap(pos + acc, acc);
+                acc + 1
+            } else {
+                acc
+            }
+        });
```

Comment:
It's not entirely obvious to me what this is doing or why. I also think the `filter().fold()` (with another fairly long chained iterator expression inside) is a great way to express this (as opposed to a syntactical loop).

Reply:
This is to solve the issue where next Client Hellos need to include the extensions in the same order as the initial one. It was the part of the previous PRs that was failing in the BoGo test as all it had been done was naively shuffling everything.

I read in the TLS spec that each extension must only appear once, so with that assumption if we saved the extension type in `input.hello.sent_extensions` before, now we can generate it in the same order.

So for example: 
```
1st Client Hello sent this way (giving numbers to ext types): |2|1|4|3|PSK|
```
Now if we generate a ClientHello like this `|4|3|1|5|6|PSK|`
It iterates through the save ClientHello and finds position of the element in the newly generated extension list:
So for example 1st iteration, it'd see that ext n2 is missing, so `acc` keeps being 0, ignores it.
2nd iteration: ext n1 is in position 2, so swap 2 with 0, `acc` is now 1 and vector in this state:
```
`|1|3|4|5|6|PSK|`
```
3rd iteration: ext n4 is in position 1 now (we skipped the 1st element as it's sorted already), so swap 2 (`pos + acc`) with 1 (`acc`); `acc` is now 2 and result of swap is:
```
|1|4|3|5|6|PSK|
```
4rd iteration: ext 3 is in position 0 (as we skip the 2 first ones), so swap 2 with 2 (granted this one was unnecessary) and `acc` is now 3, resulting of swap is:
```
|1|4|3|5|6|PSK|
```
At the end of the loop `acc` would be the number of the extensions sorted so all other new extensions after those (for example a new Cookie extension) would be shuffled, except the PSK extension of course.
And it filters out the PSK one in the beginning cause there's no point processing that as it needs to be last.

This could be done with a for statement but I kinda needed the point where the extensions stopped being sorted as per the saved vector to know where to start shuffling which would require me to create a mutable variable, so this avoids that.

Then after this it starts randomising from `start` (3 in the previous example) to the index of the PSK extension (exclusive).

Not sure if there's an action to do with this comment, sorry couldn't understand, should I document this in the code? The `skip()` as well is not needed, it's just an optimisation in order to reduce the amount of processing to do on elements that were already processed, it could be removed if you think it makes it more confusing.

Reply:
So this code is written to do nothing/unnecessary if `sent_extensions` is empty? IMO that should be made obvious by wrapping it in an `if !input.hello.sent_extensions.is_empty() { .. }`.

I would suggest that if this is doing some kind of sorting, it would be helpful to use [`sort_by_cached_key()`](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_by_cached_key) if at all possible (seems like it should be?).

Reply:
Yeah was having a look at that and it's indeed possible, does it look a bit better now?